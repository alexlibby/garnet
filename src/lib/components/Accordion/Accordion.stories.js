import Accordion from "../Accordion/Accordion.svelte";
import { BADGE } from "storybook-addon-badges";

const BADGES = {
  ...BADGE,
  ALPHA: "Alpha",
};

export default {
  title: "Garnet UI Library/Action Components/Accordion",
  component: Accordion,
  argTypes: {
    isOpen: false,
    data: [],
  },
  parameters: {
    docs: {
      story: {
        height: "500px",
      },
    },
  },
};

export const Default = {
  Component: Accordion,
  args: {
    isOpen: false,
    data: [
      {
        title: "Heading 1",
        text: "lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed malesuada, nulla sed lacinia accumsan, ligula arcu interdum urna, eget rhoncus sapien orci scelerisque metus.",
      },
      {
        title: "Heading 2",
        text: "In bibendum commodo orci nec semper. Nam magna mauris, ornare eu semper sit amet, vehicula sed metus",
      },
      {
        title: "Heading 3",
        text: "Mauris tortor mi, scelerisque nec metus nec, finibus euismod lacus. Maecenas non porttitor arcu",
      },
    ],
  },
};

Default.parameters = {
  badges: [BADGES.ALPHA],
};
