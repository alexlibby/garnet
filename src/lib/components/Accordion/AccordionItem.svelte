<svelte:options customElement="garnet-accordionitem" />

<script>
  import { slide } from "svelte/transition";
  import accordionData from "./accordiondata.json";
  export const data = accordionData;
  export let title = "";
  export let isOpen = false;
  
  const toggle = () => (isOpen = !isOpen);
</script>

<button on:click={toggle} aria-expanded={isOpen} class="accordionItem">
  <svg
    width="20"
    height="20"
    fill="none"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    stroke="currentColor"
    ><path d="M9 5l7 7-7 7" />
  </svg>
  {title}
</button>

{#if isOpen}
  <ul transition:slide={{ duration: 300 }} class="garnet">
    <slot></slot>
  </ul>
{/if}

<style>
  .garnet {
    display: flex;
    font-family: Arial, Helvetica, sans-serif;
  }

  svg {
    transition: transform 0.2s ease-in;
  }

  [aria-expanded="true"] svg {
    transform: rotate(0.25turn);
  }

  button.accordionItem {
    display: flex;
    align-items: center;
    background-color: #733635;
    color: #ffffff;
    border: none;
  }

  button[aria-expanded="false"].accordionItem {
    margin-bottom: 2px;
  }

  button.accordionItem:hover {
    background-color: #733635;
  }

  ul {
    border: 1px solid #d19c9b;
    margin: 0;
    margin-bottom: 2px;
    padding: 5px 10px 10px 20px;
  }
</style>
