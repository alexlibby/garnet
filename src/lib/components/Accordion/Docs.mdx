import { Canvas, Meta } from "@storybook/addon-docs";
import { ArgTypes } from "@storybook/blocks";
import * as AccordionStories from "./Accordion.stories";

<Meta of={AccordionStories} />

# Accordion

Button is the primary component. It has one possible state.

- [Default](#default)

## Default

This is the default state for the Accordion component:

<Canvas of={AccordionStories.Default} />

## Properties of component

Below is a list of arguments available for this component:

<ArgTypes of={AccordionStories} />
