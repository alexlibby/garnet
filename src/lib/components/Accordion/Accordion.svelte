<svelte:options customElement="garnet-accordion" />

<script>
	import AccordionItem from "./AccordionItem.svelte";
	export let data = [];	
	export let isOpen = false;
</script>

<div class="garnet-accordion">
	{#each data as entry}
		<AccordionItem title={entry.title} isOpen={isOpen}>
			<p>{entry.text}</p>
		</AccordionItem>
	{/each}
</div>

<style>
	.garnet-accordion {
		display: flex;
		flex-direction: column;
		width: 450px;
	}
</style>
