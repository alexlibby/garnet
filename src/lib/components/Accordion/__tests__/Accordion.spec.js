import { describe, test, expect } from "vitest";
import { axe } from "vitest-axe";
import * as AxeMatchers from "vitest-axe/matchers";
import { render, screen, fireEvent } from "@testing-library/svelte";
import Accordion from "../../Accordion/Accordion.svelte";

const AccordionHTML = `<div class="garnet-accordion s-zJ3CIJSsBGcs"><button aria-expanded="false" class="accordionItem s-3L_PBfp3fhj2"><svg width="20" height="20" fill="none" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" stroke="currentColor" class="s-3L_PBfp3fhj2"><path d="M9 5l7 7-7 7" class="s-3L_PBfp3fhj2"></path></svg> Heading 1</button> <!--<AccordionItem>--><button aria-expanded="false" class="accordionItem s-3L_PBfp3fhj2"><svg width="20" height="20" fill="none" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" stroke="currentColor" class="s-3L_PBfp3fhj2"><path d="M9 5l7 7-7 7" class="s-3L_PBfp3fhj2"></path></svg> Heading 2</button> <!--<AccordionItem>--><button aria-expanded="false" class="accordionItem s-3L_PBfp3fhj2"><svg width="20" height="20" fill="none" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" stroke="currentColor" class="s-3L_PBfp3fhj2"><path d="M9 5l7 7-7 7" class="s-3L_PBfp3fhj2"></path></svg> Heading 3</button> <!--<AccordionItem>--></div>`;
expect.extend(AxeMatchers);

describe("Tests for Accordion", () => {
  const mockText = "This is a Accordion";

  const props = {
    data: [
      { title: "Heading 1", text: "aaa" },
      { title: "Heading 2", text: "bbb" },
      { title: "Heading 3", text: "ccc" },
    ],
  };

  test("should render properly", () => {
    const result = render(Accordion, { props });
    expect(() => result).not.toThrow();
  });

  test("should have the correct title", () => {
    render(Accordion, props);

    expect(screen.queryByText(props.data[0].title)).toBeInTheDocument();
    expect(screen.queryByText(props.data[1].title)).toBeInTheDocument();
    expect(screen.queryByText(props.data[2].title)).toBeInTheDocument();
  });

  test("should have the correct text for the first AccordionItem", async () => {
    render(Accordion, props);

    const AccordionObj = await screen.getAllByRole("button")[0];
    await fireEvent.click(AccordionObj);

    const accordionText = await screen.findByText(/aaa/i);
    expect(accordionText).toBeTruthy();
  });

  test("should show aria-expanded as true when first item clicked", async () => {
    render(Accordion, props);
    const AccordionObj = await screen.getAllByRole("button")[0];
    await fireEvent.click(AccordionObj);
    expect(screen.getAllByRole("button", { expanded: true })[0]).not.toBeNull();
  });

  test("get a snapshot of component", () => {
    // Create a new container for the test and mount component
    const host = document.createElement("div");
    document.body.appendChild(host);
    const instance = new Accordion({ target: host, props: props });

    // check that all rendered OK
    expect(instance).toBeTruthy();

    // Take snapshot
    expect(host.innerHTML).toMatchSnapshot();
  });

  test("should demonstrate no issues with accessibility", async () => {
    const render = () => AccordionHTML;
    const result = render();

    const results = await axe(result, {
      rules: {
        region: { enabled: false },
      },
    });

    expect(results).toHaveNoViolations();
  });
});
