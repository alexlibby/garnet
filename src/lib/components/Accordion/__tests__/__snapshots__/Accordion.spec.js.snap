// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tests for Accordion > get a snapshot of component 1`] = `"<div class="garnet-accordion svelte-1lpwwml"><button aria-expanded="false" class="accordionItem svelte-1rv45at"><svg width="20" height="20" fill="none" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" stroke="currentColor" class="svelte-1rv45at"><path d="M9 5l7 7-7 7"></path></svg> Heading 1</button> <button aria-expanded="false" class="accordionItem svelte-1rv45at"><svg width="20" height="20" fill="none" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" stroke="currentColor" class="svelte-1rv45at"><path d="M9 5l7 7-7 7"></path></svg> Heading 2</button> <button aria-expanded="false" class="accordionItem svelte-1rv45at"><svg width="20" height="20" fill="none" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" viewBox="0 0 24 24" stroke="currentColor" class="svelte-1rv45at"><path d="M9 5l7 7-7 7"></path></svg> Heading 3</button> </div>"`;
